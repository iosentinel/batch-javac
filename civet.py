#!/usr/bin/python

import argparse
import sys
import tempfile
import subprocess
import shutil
import os,sys

CIVET_HOME = os.path.dirname(__file__)

if __name__ == '__main__':
  # create the parser
  parser = argparse.ArgumentParser(
      description='Civet: Java Hybrid Partial Evaluator')

  parser.add_argument(
      '-classpath', '-cp', 
      help='Specify where to find user class files and libraries')
  parser.add_argument(
      '-srcpath', '-sp', help='Specify where to find user source files')
  parser.add_argument(
      '-outsourcepath', '-s', 
      help='Specify where to place generated source files')
  parser.add_argument(
      'inputfile', default=sys.stdin,
      help='The input file for hybrid partial evaluation')

  # parse the command line  
  args = parser.parse_args()

  MAIN='com.sun.tools.javac.Main'
  p_cmd = ['java', '-ea', 
      '-classpath', args.classpath+':'+CIVET_HOME+'/dist/lib/classes.jar'+
      ':'+CIVET_HOME+'/lib/batches.jar', 
      MAIN, '-civet', 
      '-cp', args.srcpath+':'+CIVET_HOME+'/dist/lib/classes.jar',
      '-s', args.outsourcepath, args.inputfile]
  subprocess.call(p_cmd)

