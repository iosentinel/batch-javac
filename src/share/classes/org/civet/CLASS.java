package org.civet;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import sun.reflect.generics.reflectiveObjects.NotImplementedException;

import com.sun.tools.javac.code.Type;
import com.sun.tools.javac.code.Symbol.ClassSymbol;
import com.sun.tools.javac.code.Type.ArrayType;
import com.sun.tools.javac.util.List;
import com.sun.tools.javac.util.Name;

public class CLASS extends HPEObject {
  private ClassSymbol classSymbol;
  private Class<?> clazz;

  public Type getType() {
    if (isArray)
      return arrayType;
    else
      return classSymbol.type;
  }

  private ArrayType arrayType;

  public CLASS(ClassSymbol classSymbol) {
    this.kind = Kind.CLASS;
    this.classSymbol = classSymbol;
    clazz = HPEUtil.toClass(classSymbol.type);
    this.object = clazz;
  }

  public CLASS(ArrayType type) {
    this.kind = Kind.CLASS;
    isArray = true;
    arrayType = type;
  }

  @Override
  public HPEObject getFieldValue(Name fieldName) {
    if (fieldName.equals("class")) {
      return this;
    } else {
      try {
        Field f = hasField(clazz, fieldName.toString());
        if (f == null) {
          return null;
        }
        f.setAccessible(true);

        Object val = f.get(null);
        return HPEObject.fromObject(val).setTag(this.tag)
            .convertToField(fieldName, this);
      } catch (SecurityException | IllegalArgumentException
          | IllegalAccessException e) {
        throw new HPEException(e);
      }
    }
  }

  @Override
  public HPEObject clone() throws CloneNotSupportedException {
    throw new NotImplementedException();
  }

  @Override
  public String toString() {
    return classSymbol.toString();
  }

  public HPEObject newInstance(Type constructorType, List<HPEObject> args) {
    Class<?>[] paramClasses = new Class<?>[args.length()];
    try {
      int i = 0;
      Object[] argObjects = new Object[args.length()];
      for (Type t : constructorType.getParameterTypes()) {
        paramClasses[i] = HPEUtil.toClass(t);
        argObjects[i] = args.get(i).object;
        i++;
      }
      Constructor<?> constructor = clazz.getDeclaredConstructor(paramClasses);
      constructor.setAccessible(true);
      Object o = constructor.newInstance(argObjects);
      return HPEObject.fromObject(o);
    } catch (NoSuchMethodException | SecurityException | InstantiationException
        | IllegalAccessException | IllegalArgumentException
        | InvocationTargetException e) {
      throw new HPEException(e);
    }
  }

}