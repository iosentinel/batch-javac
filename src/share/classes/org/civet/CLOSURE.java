package org.civet;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

import javax.lang.model.element.ElementKind;
import javax.lang.model.type.TypeKind;

import com.sun.tools.javac.code.Type;
import com.sun.tools.javac.code.Symbol.ClassSymbol;
import com.sun.tools.javac.code.Symbol.MethodSymbol;
import com.sun.tools.javac.code.Type.MethodType;
import com.sun.tools.javac.code.Type.TypeVar;
import com.sun.tools.javac.util.List;

public class CLOSURE extends HPEObject {
  HPEObject thiz;
  MethodSymbol methodSymbol;

  public CLOSURE(HPEObject o, MethodSymbol ms) {
    thiz = o;
    methodSymbol = ms;
    this.kind = Kind.CLOSURE;
  }

  public CLOSURE(MethodSymbol ms) {
    this(null, ms);
  }

  public boolean isCTMethod() {
    assert (methodSymbol != null && methodSymbol.owner != null);
    return methodSymbol.name.toString().equals("CT")
        && methodSymbol.owner.getKind() == ElementKind.CLASS
        && ((ClassSymbol) methodSymbol.owner).fullname.toString().equals(
            "org.civet.Civet") && methodSymbol.getParameters().length() == 1;
  }

  public boolean isCTIfMethod() {
    assert (methodSymbol != null && methodSymbol.owner != null);
    return methodSymbol.name.toString().equals("CT")
        && methodSymbol.owner.getKind() == ElementKind.CLASS
        && ((ClassSymbol) methodSymbol.owner).fullname.toString().equals(
            "org.civet.Civet") && methodSymbol.getParameters().length() == 2;
  }

  public boolean isIsCTMethod() {
    assert (methodSymbol != null && methodSymbol.owner != null);
    return methodSymbol.name.toString().equals("IsCT")
        && methodSymbol.owner.getKind() == ElementKind.CLASS
        && ((ClassSymbol) methodSymbol.owner).fullname.toString().equals(
            "org.civet.Civet") && methodSymbol.getParameters().length() == 1;
  }

  public boolean isRTMethod() {
    assert (methodSymbol != null && methodSymbol.owner != null);
    return methodSymbol.name.toString().equals("RT")
        && methodSymbol.owner.getKind() == ElementKind.CLASS
        && ((ClassSymbol) methodSymbol.owner).fullname.toString().equals(
            "org.civet.Civet");
  }

  @Override
  public String toString() {
    return methodSymbol.toString();
  }

  public HPEObject invoke(List<HPEObject> args) {
    Object thisObject = null;
    if (!methodSymbol.isStatic()) {
      thisObject = thiz.object;
    }
    ClassSymbol classSymbol = (ClassSymbol) methodSymbol.owner;
    try {
      Class<?> clazz = Class.forName(classSymbol.getQualifiedName().toString());
      MethodType methodType = (MethodType) methodSymbol.type;
      Class<?>[] paramClasses = new Class<?>[args.length()];
      Object[] argObjects = new Object[args.length()];
      int i = 0;
      for (Type t : methodType.getParameterTypes()) {
        Class<?> c;
        if (t.getKind() == TypeKind.TYPEVAR) {
          c = HPEUtil.toClass(((TypeVar) t).bound);
        } else {
          c = HPEUtil.toClass(t);
        }
        paramClasses[i] = c;
        argObjects[i] = args.get(i).object;
        i++;
      }
      Method m = clazz.getDeclaredMethod(methodSymbol.name.toString(),
          paramClasses);
      m.setAccessible(true);
      Object result = m.invoke(thisObject, argObjects);
      return fromObject(result);
    } catch (ClassNotFoundException | NoSuchMethodException | SecurityException
        | IllegalAccessException | IllegalArgumentException
        | InvocationTargetException e) {
      throw new HPEException(e);
    }
  }
}