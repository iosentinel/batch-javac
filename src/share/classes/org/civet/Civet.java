package org.civet;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

public class Civet {

  public static <T> T CT(T t) {
    return t;
  }

  public static int CT(int t) {
    return t;
  }

  public static short CT(short t) {
    return t;
  }

  public static double CT(double t) {
    return t;
  }

  public static long CT(long t) {
    return t;
  }

  public static byte CT(byte t) {
    return t;
  }

  public static char CT(char t) {
    return t;
  }

  public static boolean CT(boolean t) {
    return t;
  }

  public static float CT(float t) {
    return t;
  }

  /////////////////
  public static <T> T CT(T t, Boolean b) {
    return t;
  }

  public static int CT(int t, Boolean b) {
    return t;
  }

  public static short CT(short t, Boolean b) {
    return t;
  }

  public static double CT(double t, Boolean b) {
    return t;
  }

  public static long CT(long t, Boolean b) {
    return t;
  }

  public static byte CT(byte t, Boolean b) {
    return t;
  }

  public static char CT(char t, Boolean b) {
    return t;
  }

  public static boolean CT(boolean t, Boolean b) {
    return t;
  }

  public static float CT(float t, Boolean b) {
    return t;
  }
  
  /////////////////
  
  public static <T> T RT(T t) {
    return t;
  }

  public static int RT(int t) {
    return t;
  }

  public static short RT(short t) {
    return t;
  }

  public static double RT(double t) {
    return t;
  }

  public static long RT(long t) {
    return t;
  }

  public static byte RT(byte t) {
    return t;
  }

  public static char RT(char t) {
    return t;
  }

  public static boolean RT(boolean t) {
    return t;
  }

  public static float RT(float t) {
    return t;
  }

  public static <T> Boolean IsCT(T o) {
    return Boolean.TRUE;
  }

  public static Boolean IsCT(byte o) {
    return Boolean.TRUE;
  }

  public static Boolean IsCT(char o) {
    return Boolean.TRUE;
  }

  public static Boolean IsCT(short o) {
    return Boolean.TRUE;
  }

  public static Boolean IsCT(int o) {
    return Boolean.TRUE;
  }

  public static Boolean IsCT(long o) {
    return Boolean.TRUE;
  }

  public static Boolean IsCT(float o) {
    return Boolean.TRUE;
  }

  public static Boolean IsCT(double o) {
    return Boolean.TRUE;
  }

  public static Boolean IsCT(boolean o) {
    return Boolean.TRUE;
  }

  @Retention(RetentionPolicy.CLASS)
  @Target(ElementType.METHOD)
  public static @interface Compile {
    
  }
}
