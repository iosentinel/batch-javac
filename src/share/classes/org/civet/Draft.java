package org.civet;

public class Draft {

  
  enum Test { A1, A2, A3 };
  
  
  /**
   * @param args
   */
  public static void main(String[] args) {
    System.out.println(Test.class.isEnum());
    for (Test x : Test.class.getEnumConstants()) {
      System.out.println(x);
    }
  }

}
