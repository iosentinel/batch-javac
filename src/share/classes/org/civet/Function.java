package org.civet;

import com.sun.tools.javac.util.List;
import com.sun.tools.javac.util.ListBuffer;

public abstract class Function<R, T> {
  public abstract R apply(T t);

  public static <S, P> List<S> map(Function<S, P> f, List<? extends P> l) {
    if (l == null)
      return null;
    ListBuffer<S> r = new ListBuffer<>();
    for (P t : l) {
      r.append(f.apply(t));
    }
    return r.toList();
  }

  public static <S, P> List<S> map(Function<S, P> f,
      Function<Boolean, P> filter, List<? extends P> l) {
    if (l == null)
      return null;
    ListBuffer<S> r = new ListBuffer<>();
    for (P t : l) {
      if (!filter.apply(t))
        r.append(f.apply(t));
    }
    return r.toList();
  }

  public static <S> List<S> filter(Function<Boolean, S> f, List<? extends S> l) {
    if (l == null)
      return null;
    ListBuffer<S> r = new ListBuffer<>();
    for (S t : l) {
      if (!f.apply(t))
        r.append(t);
    }
    return r.toList();
  }
}
