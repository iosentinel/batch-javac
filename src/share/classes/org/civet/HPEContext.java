package org.civet;

import java.util.HashSet;
import java.util.Set;

import javax.lang.model.element.ElementKind;
import javax.lang.model.type.TypeKind;

import com.sun.source.tree.MethodInvocationTree;
import com.sun.source.tree.Tree.Kind;
import com.sun.source.util.Trees;
import com.sun.tools.javac.api.JavacTrees;
import com.sun.tools.javac.code.Flags;
import com.sun.tools.javac.code.Symbol;
import com.sun.tools.javac.code.Symbol.ClassSymbol;
import com.sun.tools.javac.code.Symbol.MethodSymbol;
import com.sun.tools.javac.code.Symbol.TypeSymbol;
import com.sun.tools.javac.code.Symbol.VarSymbol;
import com.sun.tools.javac.code.Symtab;
import com.sun.tools.javac.code.Type;
import com.sun.tools.javac.code.Type.MethodType;
import com.sun.tools.javac.code.Types;
import com.sun.tools.javac.jvm.ClassReader.SourceCompleter;
import com.sun.tools.javac.tree.JCTree.JCExpression;
import com.sun.tools.javac.tree.JCTree.JCFieldAccess;
import com.sun.tools.javac.tree.JCTree.JCIdent;
import com.sun.tools.javac.tree.JCTree.JCMethodInvocation;
import com.sun.tools.javac.tree.TreeMaker;
import com.sun.tools.javac.util.Context;
import com.sun.tools.javac.util.Filter;
import com.sun.tools.javac.util.List;
import com.sun.tools.javac.util.ListBuffer;
import com.sun.tools.javac.util.Name;
import com.sun.tools.javac.util.Names;
import com.sun.tools.javac.util.Pair;

public class HPEContext {
  public HPEEnvironment env;
  public Context javacContext;
  public TreeMaker treeMaker;
  public Types types;
  public Names names;
  public Trees trees;
  public SourceCompleter sourceCompleter;
  public Symtab symtab;

  public HPEContext(HPEEnvironment env, Context javacContext) {
    super();
    this.env = env;
    this.javacContext = javacContext;
    this.treeMaker = TreeMaker.instance(javacContext);
    this.types = Types.instance(javacContext);
    this.names = Names.instance(javacContext);
    this.trees = JavacTrees.instance(javacContext);
    this.symtab = Symtab.instance(javacContext);
  }

  private List<Method> genMethods = List.<Method> nil();
  private Set<String> methodNames = new HashSet<String>();

  Pair<Method, Boolean> generateName(MethodSymbol methodSymbol,
      List<Parameter> staticArgs) {

    // Look if a name for this call has already been generated.
    Method m = new Method(methodSymbol, staticArgs, this);
    int i = genMethods.indexOf(m);

    if (i >= 0) // Already generated.
      return Pair.of(genMethods.get(i), true);

    // Generate a new name.
    String genName = "";
    do {
      if (methodSymbol.getKind() == ElementKind.CONSTRUCTOR) {
        genName = methodSymbol.owner.name.toString() + "$" + getNextNameSeqNum();
      }
      else {
        genName = methodSymbol.name.toString() + "$" + getNextNameSeqNum();
      }
    } while (methodNames.contains(genName));

    methodNames.add(genName);
    genMethods = genMethods.append(m.setGenMethodName(genName));

    return Pair.of(m, false);
  }

  private static Integer nameSeqNum = new Integer(1000);

  private static String getNextNameSeqNum() {
    nameSeqNum++;
    return String.valueOf(nameSeqNum);
  }

  boolean equalsMethodSym(MethodSymbol m1, MethodSymbol m2) {
    return m1.name.toString().equals(m2.name.toString())
        && this.types.isSameType(m1.type, m2.type)
        && (this.types.isSameType(m1.owner.type, m2.owner.type));
  }

  boolean equalsMethodSignature(MethodSymbol m1, MethodSymbol m2) {
    return m1.name.toString().equals(m2.name.toString())
        && this.types.isSameType(m1.type, m2.type);
  }

  MethodSymbol retMethodSymbol(MethodInvocationTree node) {
    JCMethodInvocation mi = (JCMethodInvocation) node;
    if (mi.meth.getKind() == Kind.MEMBER_SELECT) {
      JCFieldAccess methAccess = (JCFieldAccess) mi.meth;
      return (MethodSymbol) methAccess.sym;
    } else if (mi.meth.getKind() == Kind.IDENTIFIER) {
      JCIdent methAccess = (JCIdent) mi.meth;
      return (MethodSymbol) methAccess.sym;
    }
    return null;
  }

  MethodSymbol generateMethodSymbol(MethodSymbol ms, Name newName,
      List<Type> rtTypes, List<Parameter> ctParams, boolean isStatic) {
    assert (ms.owner.getKind() == ElementKind.CLASS || ms.owner.getKind() == ElementKind.INTERFACE);
    MethodType mt = new MethodType(rtTypes, ms.getReturnType(),
        ms.getThrownTypes(), ms.type.tsym);
    long flags = isStatic ? ms.flags_field | Flags.STATIC : ms.flags_field;
    MethodSymbol genMS = new MethodSymbol(flags, newName, mt, ms.owner);
    genMS.params = List.<VarSymbol> nil();
    for (int i = 0; i < ms.params.length(); i++) {
      VarSymbol pv = ms.params.get(i);
      boolean isCT = false;
      for (Parameter p : ctParams) {
        if (p.position == i) {
          isCT = true;
          break;
        }
      }
      if (!isCT) {
        genMS.params = genMS.params.append(pv);
      }
    }
    return genMS;
  }

  JCMethodInvocation generateStaticMethodInvocation(MethodSymbol ms,
      List<JCExpression> args) {
    assert (ms.params.length() == args.length());
    // ClassSymbol clazz = (ClassSymbol) ms.owner;
    JCMethodInvocation mi = treeMaker.App(
        treeMaker.Select(treeMaker.Ident(ms.owner.getQualifiedName()), ms),
        args);
    return mi;
  }

  JCMethodInvocation generateInstanceMethodInvocation(JCExpression selected,
      MethodSymbol ms, List<JCExpression> args) {
    JCMethodInvocation mi = treeMaker.App(treeMaker.Select(selected, ms), args);
    return mi;
  }

  JCMethodInvocation generateInstanceMethodInvocation(MethodSymbol ms,
      List<JCExpression> args) {
    JCMethodInvocation mi = treeMaker.App(treeMaker.Ident(ms), args);
    return mi;
  }

  boolean hasSource(Symbol s) {
    return trees.getTree(s) != null;
  }

  List<ClassSymbol> subTypes(Symbol c) {
    ListBuffer<ClassSymbol> subs = new ListBuffer<>();
    ClassSymbol[] classes = symtab.classes.values().toArray(
        new ClassSymbol[] {});
    for (int i = 0; i < classes.length; i++) {
      if (hasSource(classes[i]) && classes[i].isSubClass(c, types)) {
        subs.append(classes[i]);
      }
    }
    return subs.toList();
  }

  TypeSymbol findSymbolForClass(Class<?> clazz) {
    String className = clazz.getName();
    if (clazz.isPrimitive()) {
      switch (clazz.getName()) {
        case "int":
          return findSymbolForClass(Integer.class);
        case "short":
          return findSymbolForClass(Short.class);
        case "long":
          return findSymbolForClass(Long.class);
        case "float":
          return findSymbolForClass(Float.class);
        case "double":
          return findSymbolForClass(Double.class);
        case "byte":
          return findSymbolForClass(Byte.class);
        case "char":
          return findSymbolForClass(Character.class);
        default:
          throw new HPEException("not implemented for " + clazz);
      }
    }
    return symtab.classes.get(names.fromString(className));
  }

  MethodSymbol findMethodSymbolInClassSymbol(final MethodSymbol m, ClassSymbol c) {
    Iterable<Symbol> elems = c.members().getElements(new Filter<Symbol>() {
      @Override
      public boolean accepts(Symbol t) {
        return t instanceof MethodSymbol
            && t.name.toString().equals(m.name.toString())
            && HPEContext.this.types.isSameType(t.type, m.type);
      }
    });
    if (elems.iterator().hasNext()) {
      return (MethodSymbol) elems.iterator().next();
    } else {
      if (c.getSuperclass().getKind() == TypeKind.NONE) {
        return m;
      } else {
        return findMethodSymbolInClassSymbol(m,
            (ClassSymbol) c.getSuperclass().tsym);
      }
    }
  }

  Type getType(Class<?> clazz) {
    if (clazz.isArray()) {
      return new Type.ArrayType(getType(clazz.getComponentType()),
          symtab.arrayClass);
    }
    return findSymbolForClass(clazz).type;
  }

}
