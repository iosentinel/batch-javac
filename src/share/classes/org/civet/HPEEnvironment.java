package org.civet;

import java.util.Iterator;
import java.util.Stack;

import com.sun.tools.javac.code.Symbol;


public class HPEEnvironment {
  /**
   * This is a stack of method calls.
   */
  private Stack<HPEStack> stack = new Stack<HPEStack>();

  public HPEObject getCurrentThis() {
    return getCurrentStack().thisObject;
  }

  public HPEEnvironment() {

  }

  /**
   * create a new environment with the heap and the given PE stack.
   *
   * @param heap
   * @param peStack
   */
  public HPEEnvironment(HPEStack peStack) {
    super();
    this.stack.push(peStack);
  }

  /**
   * this method should be called upon calling a method. it creates a new stack
   * and push it into environment.
   */
  public void inStack(HPEObject thiz) {
    stack.push(new HPEStack(thiz));
  }

  /**
   * this method clone the last method call stack and return a new environment
   * with the current heap. This is useful when dealing with branches.
   */
  public HPEEnvironment cloneCurrentStackEnv() {
    return new HPEEnvironment(stack.peek().clone());
  }

  /**
   * This method should be called upon returning from a method call.
   */
  public void outStack() {
    stack.pop();
  }

  /**
   * This method should be called upon going into a new block. It calls the
   * inFrame method from the current stack.
   */
  public void inFrame() {
    getCurrentStack().inFrame();
  }

  /**
   * This method should be called upon exiting a block.
   */
  public void outFrame() {
    getCurrentStack().outFrame();
  }

  /**
   *
   * @return Returns the current stack
   */
  public HPEStack getCurrentStack() {
    return stack.peek();
  }

  public HPEObject lookup(HPESymbol n) {
    HPEObject e = stack.peek().lookup(n);
    return e;
  }

  public boolean contains(HPESymbol n) {
    return stack.peek().contains(n);
  }

  public void add(HPESymbol n, HPEObject v) {
    if (n.isField) {
      n.owner.setFieldValue(n.name, v);
    }
    else stack.peek().add(n, v);
    
  }

  public void remove(Symbol n) {
    throw new Error("should not be in remove");
  }

  public String toString() {
    Iterator<HPEStack> i = stack.iterator();
    StringBuffer s = new StringBuffer("(\n");
    while (i.hasNext()) {
      s.append(i.next() + ",\n");
    }
    s.append(")");
    return s.toString();
  }

  public void removeAltered(HPEEnvironment env) {
    // this.getCurrentStack().removeAltered(env.getCurrentStack());
    throw new Error("should not be in removeAltered");
  }
}
