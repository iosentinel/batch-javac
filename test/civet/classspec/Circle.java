package classspec;

import org.civet.Civet;
import org.civet.Civet.Compile;

public class Circle {

  int radius;
  Point center;

  public Circle() {
  }

  public Circle(int radius, Point center) {
    this.radius = radius;
    this.center = center;
  }

  @Compile
  public static void main(String[] args) {
    Circle c1 = new Circle(Civet.CT(10), new Point(1, 1));
    System.out.println(c1.getArea());
  }

  double getArea() {
    return Math.PI * radius * radius;
  }

  double getPerimeter() {
    return 2 * Math.PI * radius;
  }

}

class Point {

  double x, y;

  public Point(double x, double y) {
    super();
    this.x = x;
    this.y = y;
  }

}