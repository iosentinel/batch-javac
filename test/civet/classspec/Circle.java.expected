package classspec;

import org.civet.Civet;
import org.civet.Civet.Compile;

class Circle {
    int radius;
    Point center;
    
    public Circle() {
        super();
    }
    
    public Circle(int radius, Point center) {
        super();
        this.radius = radius;
        this.center = center;
    }
    
    @Compile()
    public static void main(String[] args) {
        Circle c1 = new Circle$1001(new Point(1, 1));
        System.out.println(c1.getArea());
    }
    
    double getArea() {
        return Math.PI * radius * radius;
    }
    
    double getPerimeter() {
        return 2 * Math.PI * radius;
    }
}
class Point {
    double x;
    double y;
    
    public Point(double x, double y) {
        super();
        this.x = x;
        this.y = y;
    }
}
class Circle$1001 extends Circle {
    
    public Circle$1001(classspec.Point center) {
        super();
        this.center = center;
    }
    
    double getArea() {
        return 314.1592653589793;
    }
    
    double getPerimeter() {
        return 62.83185307179586;
    }
}
